components/script/dom/htmlscriptelement.rs

```diff
@@ -53,6 +58,32 @@ use std::sync::{Arc, Mutex};
 use style::str::{StaticStringVec, HTML_SPACE_CHARACTERS};
 use uuid::Uuid;
 
+pub struct OffThreadCompilationContext {
+    script_element: Trusted<HTMLScriptElement>,
+    script_origin: ScriptOrigin,
+    script_kind: ExternalScriptKind,
+    task_source: NetworkingTaskSource,
+    canceller: TaskCanceller,
+}
+
+#[allow(unsafe_code)]
+unsafe extern "C" fn off_thread_compilation_callback(
+    token: *mut OffThreadToken,
+    callback_data: *mut c_void,
+) {
+    let off_thread_compilation_context: &mut OffThreadCompilationContext =
+        &mut *(callback_data as *mut OffThreadCompilationContext);
+    let elem = off_thread_compilation_context.script_element.root();
+    let global = elem.global();
+    let _ = off_thread_compilation_context.task_source.queue_with_canceller(
+        task! (off_thread_compile_continue: move || {
+            let global_context = global;
+            off_thread_compilation_context.script_origin.code = SourceCode::Compiled(*FinishOffThreadScript(*global_context.get_cx(), token));
```

Comment:
So I think we're going to have to do two things:

1. Root the `JsScript` as soon as it is received here, using the `rooted!` macro(you'll see various example across the codebase, just search you'll find many). 
2. store it inside a `Heap` inside the `ScriptOrigin`. For that I think you'll have to do something like:

```rust
let rooted_script = rooted! //
let heap = Heap::default();
heap.set(rooted_script.get());
let source_code = SourceCode::Compiled(heap);
```

Reply:
What is the significance of rooted?

Reply:
https://github.com/servo/servo/blob/6b0d9afd6fdc28356ad44af0104ddd25a7b6438d/components/script/dom/bindings/root.rs#L5

https://developer.mozilla.org/en-US/docs/Mozilla/Projects/SpiderMonkey/GC_Rooting_Guide